src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs

```diff
@@ -220,13 +223,10 @@ public virtual int Read()
             return _singleChar[0];
         }
 
-        public virtual bool ReadBoolean()
-        {
-            FillBuffer(1);
-            return (_buffer[0] != 0);
-        }
+        public virtual byte ReadByte() => InternalReadByte();
 
-        public virtual byte ReadByte()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private byte InternalReadByte()
```

Comment:
Does this implementation end up being faster than just having ReadByte be `=> InternalRead(1)[0]`?

Reply:
And this is also a potential breaking change because of `Stream.ReadByte()` is virtual. It won't be surprising if somebody took dependency on `BinaryReader.ReadByte()` calling `Stream.ReadByte()`.

Reply:
Both original and my new implementation is calling Stream.ReadByte() from the Binaryreader.ReadByte() so I am not sure I understand this comment. I don't think I changed anything there. I did change ReadBoolean and ReadSByte to use the Strea,ReadByte(). I think it is more correct.

Reply:
Calling ReadByte() should be definitely faster for properly implemented streams. 

Calling ReadByte() instead of Read() is not bug-for-bug compatible and it can break for incorrectly implemented streams. I think it is ok to do this change in .NET Core. @stephentoub Do you agree?

Reply:
The only 2 functions I changed is ReadBoolean and ReadSByte. As ReadByte already called the Stream.ReadByte than I think the change here does make sense. If streams didn't implement ReadByte correctly than the BinaryReader.ReadByte was already broken even before my change.

Reply:
> Calling ReadByte() instead of Read() is not bug-for-bug compatible and it can break for incorrectly implemented streams. I think it is ok to do this change in .NET Core. @stephentoub Do you agree?

From a compatibility standpoint I think it's ok.  And as you say, for optimized streams it should be better.  My concern is only that some streams don't override ReadByte, in which case they pick up the base implementation, which allocates a new byte[1] for each call, which will make it worse.  On the balance, it's probably a fine change to make: all of our streams override ReadByte, and any stream that cared about performance would as well.